### DISCUSSION TITLE
how to list treesitter symbols in mini.pick

### DISCUSSION THREAD

**Discussion by dhananjaylatkar** [nvim-mini/mini.nvim#1220] [Q&A] [answered]
I'm looking for alternative to `:Telescope treesitter symbols=function`

**Comment by echasnovski** [comment, marked as answer]
Apart from built-in pickers in 'mini.pick' itself, there are [additional pickers in 'mini.extra'](https://github.com/echasnovski/mini.nvim/blob/e50cf9de614500a20e47cfc50e30a100042f91c3/doc/mini-extra.txt#L211). In particular, there is a [`treesitter`](https://github.com/echasnovski/mini.nvim/blob/e50cf9de614500a20e47cfc50e30a100042f91c3/doc/mini-extra.txt#L722) picker.

So after calling `require('mini.pick').setup()` and `require('mini.extra').setup()`, you can execute `:Pick treesitter` and then type `'function` to filter nodes that contain "function" in their name.

**Comment by dhananjaylatkar** [comment]
I shows up like this -

![image](https://github.com/user-attachments/assets/f8a2aa7f-5a93-49d4-a525-226e4506c089)

How to see function names without going in preview mode

**Comment by dhananjaylatkar** [comment]
Created picker to show given type of TS symbols and preview node text
```lua
M.treesitter = function(type)
  local buf_id = vim.api.nvim_get_current_buf()
  local has_parser, parser = pcall(vim.treesitter.get_parser, buf_id)
  if not has_parser or parser == nil then
    return
  end

  -- Make items by traversing roots of all trees (including injections)
  local items, traverse = {}, nil
  traverse = function(node, depth)
    if depth >= 1000 then
      return
    end
    for child in node:iter_children() do
      if child:named() then
        if child:type() == type then
          local lnum, col, end_lnum, end_col = child:range()
          lnum, col, end_lnum, end_col = lnum + 1, col + 1, end_lnum + 1, end_col + 1
          local text = vim.treesitter.get_node_text(child, buf_id)
          local item = { text = text, bufnr = buf_id, lnum = lnum, col = col, end_lnum = end_lnum, end_col = end_col }
          table.insert(items, item)
        end

        traverse(child, depth + 1)
      end
    end
  end

  parser:for_each_tree(function(ts_tree, _)
    traverse(ts_tree:root(), 0)
  end)

  return M.pick_start(items, { source = { name = "Tree-sitter nodes" } })
end
```

![image](https://github.com/user-attachments/assets/10d220ef-a360-4502-8fd2-82e191581e79)

**Reply by echasnovski** [reply]
Nice look! Even better if it works as you want it to.
